### ISSUE TITLE
Newbie question: when do caches get persisted?

### ISSUE THREAD

**Issue by xstos** [microsoft/FASTER#466]
Should a log be created when running the quick-start example?

https://microsoft.github.io/FASTER/docs/quick-start-guide/

I ran it and I don't see any files produced. What gives? I tried looking through the source code to figure out when/where files are created but didn't get very far.

Sorry for this incredible basic question.

**Comment by badrishc** [comment]
A log on disk does not need to be created until you either (1) run out of memory space; or (2) take a checkpoint to explicitly persist the whole log.

When creating FASTER, you defined the size of memory part of the log via `LogSettings`. Until that overflows, there is no reason for FASTER to write to disk.

When you need to checkpoint, you run this: `await store.TakeHybridLogCheckpointAsync(CheckpointType.FoldOver)`.

**Comment by xstos** [comment]
Very cool, thanks! I added a `store.Recover()` upon init and the checkpoint loads great. I noticed during my experiments that when I upsert/overwrite the same data, the log still grows after checkpoints. The solution that comes to mind to "compact" the cache would be to simply create a new FasterKV pointing to a different file and `store.Iterate` & copy over the existing data? Is there a built-in mechanism for this? Not super complicated regardless. Just idle curiosity :)

Much obliged.

**Comment by badrishc** [comment]
If you are often writing the same keys between checkpoints, a snapshot checkpoint (CheckpointType.Snapshot) could be superior. A fold-over checkpoint makes the log read-only so future updates are appended to the log. OTOH, a snapshot checkpoint takes a snap of main memory to a _separate_ file on disk, so that the main log will not grow. Read here for details: https://microsoft.github.io/FASTER/docs/fasterkv-recovery/

For log compaction, we actually have in-built support, see here: https://microsoft.github.io/FASTER/docs/fasterkv-basics/#log-compaction

**Comment by xstos** [comment]
Awesome thanks! I'll close this issue and stop bugging you :)

**Comment by badrishc** [comment]
Sure! Just a minor comment in the code above, the synchronous `TakeHybridLogCheckpoint` only initiates the checkpoint and it does not complete immediately. To ensure it has completed in sync code, you need to call: `store.CompleteCheckpointAsync().GetAwaiter().GetResult()`.

**Comment by xstos** [comment]
Willdo thanks! 

Forgot to mention something silly. I once tried an experiment with the stock `Dictionary<K,V>` implementation from Microsoft reference source, and hacked its innards to serialize keys/values to a Memory Mapped File using a simple protobuf-type binary serializer I concocted (that could handle arbitrary objects and used mono to create fast reflection accessors). I managed to achieve performance maybe 10X slower than a regular Dictionary with a similar use case to the above example. I think your solution blows my hackery out of the water. I've been waiting for something like this to emerge for a while now. Kudos and amazing job.
